src/layout/MasterLayout.cpp

```diff
@@ -1071,42 +1071,71 @@ std::any CHyprMasterLayout::layoutMessage(SLayoutMessageHeader header, std::stri
         recalculateMonitor(header.pWindow->m_iMonitorID);
 
     } else if (command == "orientationnext") {
-        const auto PWINDOW = header.pWindow;
-
-        if (!PWINDOW)
-            return 0;
+        runOrientationCycle(&header, nullptr, 1);
+    } else if (command == "orientationprev") {
+        runOrientationCycle(&header, nullptr, -1);
+    } else if (command == "orientationcycle") {
+        runOrientationCycle(&header, &vars, 1);
+    }
 
-        prepareLoseFocus(PWINDOW);
+    return 0;
+}
 
-        const auto PWORKSPACEDATA = getMasterWorkspaceData(PWINDOW->m_iWorkspaceID);
+// If vars is null, we use the default list
+void CHyprMasterLayout::runOrientationCycle(SLayoutMessageHeader* header, CVarList* vars, int direction) {
+    std::vector<eOrientation> cycle;
+    if (vars != nullptr)
+        buildOrientationCycleVectorFromVars(&cycle, vars);
 
-        if (PWORKSPACEDATA->orientation == ORIENTATION_CENTER) {
-            PWORKSPACEDATA->orientation = ORIENTATION_LEFT;
-        } else {
-            PWORKSPACEDATA->orientation = (eOrientation)(PWORKSPACEDATA->orientation + 1);
-        }
+    if (cycle.size() == 0)
+        buildOrientationCycleVectorFromEOperation(&cycle);
 
-        recalculateMonitor(header.pWindow->m_iMonitorID);
-    } else if (command == "orientationprev") {
-        const auto PWINDOW = header.pWindow;
+    const auto PWINDOW = header->pWindow;
 
-        if (!PWINDOW)
-            return 0;
+    if (!PWINDOW)
+        return;
 
-        prepareLoseFocus(PWINDOW);
+    prepareLoseFocus(PWINDOW);
 
-        const auto PWORKSPACEDATA = getMasterWorkspaceData(PWINDOW->m_iWorkspaceID);
+    const auto PWORKSPACEDATA = getMasterWorkspaceData(PWINDOW->m_iWorkspaceID);
 
-        if (PWORKSPACEDATA->orientation == ORIENTATION_LEFT) {
-            PWORKSPACEDATA->orientation = ORIENTATION_CENTER;
-        } else {
-            PWORKSPACEDATA->orientation = (eOrientation)(PWORKSPACEDATA->orientation - 1);
+    int        nextOrPrev = 0;
+    for (size_t i = 0; i < cycle.size(); ++i) {
+        if (PWORKSPACEDATA->orientation == cycle.at(i)) {
+            nextOrPrev = i + direction;
+            break;
         }
+    }
 
-        recalculateMonitor(header.pWindow->m_iMonitorID);
+    if (nextOrPrev >= (int)cycle.size())
```

Comment:
future-proofing this would be better. Since you already take `direction` as an int, who knows if in the future someone wont pass -3 or +1337?

```cpp
    if (nextOrPrev >= (int)cycle.size())
        nextOrPrev = nextOrPrev % (int)cycle.size();
    else if (nextOrPrev < 0)
        nextOrPrev = cycle.size() + (nextOrPrev % (int)cycle.size());
```